mentat/code_context.py

```diff
@@ -199,147 +117,306 @@ async def refresh(self, replace_paths: bool | None = False):
-                        level_message = "partial syntax tree"
-                    case "filenames":
-                        level_message = "filepaths only"
-
-                message = f"\nIncluding CodeMap ({level_message})"
-                await stream.send(message, color="green")
-                code_message += [f"\n{code_map_message.content}"]
-            else:
-                message = [
-                    "\nExcluding CodeMap from system message.",
-                    "Reason: not enough tokens available in model context.",
+        return "\n".join(code_message)
+
+    def _get_include_features(self) -> list[CodeFile]:
+        include_features = list[CodeFile]()
+        for path, feature in self.include_files.items():
+            if feature.level == CodeMessageLevel.INTERVAL:
+                interval_str = ",".join(f"{i.start}-{i.end}" for i in feature.intervals)
+                path = f"{path}:{interval_str}"
+            diff_target = (
+                self.diff_context.target if path in self.diff_context.files else None
+            )
+            feature = CodeFile(path, feature.level, diff=diff_target)
+            include_features.append(feature)
+
+        def _feature_relative_path(f: CodeFile) -> str:
+            return os.path.relpath(f.path, self.config.git_root)
+
+        return sorted(include_features, key=_feature_relative_path)
+
+    async def _get_auto_features(
+        self,
+        code_file_manager: "CodeFileManager",
+        parser: "Parser",
+        model: str,
+        include_features: list[CodeFile],
+        max_tokens: Optional[int] = None,
+    ) -> list[CodeFile]:
+        # Generate all possible permutations for all files in the project.
+        candidate_features = list[CodeFile]()
+        for path in get_non_gitignored_files(self.config.git_root):
+            if not path.is_dir() and not is_file_text_encoded(path):
+                continue
+            permutations: list[tuple[CodeMessageLevel, str | None]] = [
+                (CodeMessageLevel.CODE, None),
+                (CodeMessageLevel.FILE_NAME, None),
+            ]
+            if not self.settings.no_code_map:
+                permutations += [
+                    (CodeMessageLevel.CMAP_FULL, None),
+                    (CodeMessageLevel.CMAP, None),
+                ]
+            if self.diff_context.target and path in self.diff_context.files:
+                permutations += [
+                    (level, self.diff_context.target) for level, _ in permutations
                 ]
-                message = "\n".join(message)
-                await stream.send(message, color="yellow")
+            for level, diff in permutations:
+                feature = CodeFile(path, level=level, diff=diff)
+                candidate_features.append(feature)
 
-        return "\n".join(code_message)
+        # Sort candidates by relevance/density.
+        tokens_remaining = 1e6 if max_tokens is None else max_tokens
```

Comment:
why is `max_tokens` sometimes `None`?

Reply:
Actually it shouldn't be, I'll make it a required arg.